test/e2e/activator_test.go

```diff
@@ -0,0 +1,174 @@
+	var (
+		logger  = logging.GetContextLogger(t.Name())
+		clients = Setup(t)
+	)
+	const (
+		// The number of concurrent requests to hit the activator with.
+		// 1000 = the number concurrent connections in Istio.
+		concurrency = 1000
+		// Timeout to wait for the responses.
+		// Ideally we must wait ~30 seconds, TODO: need to figure out where the delta comes from.
+		timeout = 65 * time.Second
+		// How long the service will process the request in ms.
+		serviceSleep = 300
+	)
+
+	names := test.ResourceNames{
+		Service: test.AppendRandomString(configName),
+		Image:   "observed-concurrency",
+	}
+
+	configOptions := test.Options{
+		ContainerConcurrency: 1,
+	}
+
+	fopt := func(service *v1alpha1.Service) {
+		service.Spec.RunLatest.Configuration.RevisionTemplate.Annotations = map[string]string{"autoscaling.knative.dev/maxScale": "10"}
+	}
+
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	// Create a service with concurrency 1 that could sleep for N ms.
+	// Limit its maxScale to 10 containers, wait for the service to scale down and hit it with concurrent requests.
+	resources, err := test.CreateRunLatestServiceReady(t, clients, &names, &configOptions, fopt)
+	if err != nil {
+		t.Fatalf("Unable to create resources: %v", err)
+	}
+	domain := resources.Route.Status.Domain
+
+	deploymentName := resources.Revision.Name + "-deployment"
+
+	logger.Info("Waiting for deployment to scale to zero.")
+	if err := pkgTest.WaitForDeploymentState(
+		clients.KubeClient,
+		deploymentName,
+		test.DeploymentScaledToZeroFunc,
+		"DeploymentScaledToZero",
+		test.ServingNamespace,
+		2*time.Minute); err != nil {
+		t.Fatalf("Failed waiting for deployment to scale to zero: %v", err)
+	}
+
+	url := fmt.Sprintf("http://%s/?timeout=%d", domain, serviceSleep)
+
+	client, err := pkgTest.NewSpoofingClient(clients.KubeClient, t.Logf, domain, test.ServingFlags.ResolvableDomain)
+	client.RequestTimeout = timeout
+
+	sendRequests(client, url, concurrency, timeout, logger, t)
+
+}
+
+func sendRequests(client *spoof.SpoofingClient, url string, concurrency int, timeout time.Duration, logger *logging.BaseLogger, t *testing.T) {
+	t.Helper()
+	var (
+		group        errgroup.Group
+		responses    int32
+		wantResponse = http.StatusOK
+		resChannel   = make(chan *spoof.Response, concurrency)
+	)
+	timeoutChan := time.After(timeout)
+	errChan := make(chan error)
+
+	// Send out the requests asynchronously and wait for them to finish.
+	logger.Info("Starting to send out the requests")
+
+	// Print out stats and wait for a global timeout.
+	go func() {
+		for {
+			select {
+			case <-time.Tick(5 * time.Second):
+				logger.Infof("Received responses: %d", atomic.LoadInt32(&responses))
+				if concurrency == atomic.LoadInt32(&responses) {
+					return
+				}
+			case <-timeoutChan:
+				return
+			}
+		}
+	}()
+
+	// Send requests async and wait for the responses.
+	for i := 0; i < concurrency; i++ {
```

Comment:
I don't get the need for `resChannel` here at all.
You're generating exactly N requests and you're waiting for all requests to complete before you even start processing the responses.

So just change your responses to be a slice of responses of size `concurrency` and give each goroutine the pointer to the i-th element in the slice.

Then, after the group.Wait() you either fail the test right there, if there was an error, or you process the response list, since at this point you know all requests finished with a response.

Reply:
Thank you for the suggestion. 
I personally prefer working with channels if there is concurrency involved. In this case I also think that sharing a slice (even with assigning a separate pointer to each go routine) doesn't adhere to common practice in go. From https://golang.org/doc/effective_go.html#concurrency
> Do not communicate by sharing memory; instead, share memory by communicating.

Reply:
We don't really communicate here, but rather assign memory. But up to you.
Also channels respond to `len`, so that might be used as a proxy for `responses`, though once it starts to be consumed below it won't be correct anymore.